miniwin/src/d3drm/backends/citro3d/renderer.cpp

```diff
@@ -0,0 +1,617 @@
+			C3D_FVUnifSet(
+				GPU_VERTEX_SHADER,
+				uLoc_lightVec + 1,
+				-light.direction.x,
+				-light.direction.y,
+				-light.direction.z,
+				0.0f
+			);
+			C3D_FVUnifSet(GPU_VERTEX_SHADER, uLoc_lightClr + 1, lightColor.r, lightColor.g, lightColor.b, 0.0f);
+		}
+		else if (light.positional == 1.0f) {
+			C3D_FVUnifSet(
+				GPU_VERTEX_SHADER,
+				uLoc_lightVec + 0,
+				light.position.x,
+				light.position.y,
+				light.position.z,
+				0.0f
+			);
+			C3D_FVUnifSet(GPU_VERTEX_SHADER, uLoc_lightClr + 0, lightColor.r, lightColor.g, lightColor.b, 0.0f);
+		}
+	}
+
+	return S_OK;
+}
+
+void Citro3DRenderer::EnableTransparency()
+{
+	C3D_DepthTest(true, GPU_GREATER, GPU_WRITE_COLOR);
+}
+
+void ConvertMatrix(const D3DRMMATRIX4D in, C3D_Mtx* out)
+{
+	for (int i = 0; i < 4; i++) {
+		out->r[i].x = in[0][i];
+		out->r[i].y = in[1][i];
+		out->r[i].z = in[2][i];
+		out->r[i].w = in[3][i];
+	}
+}
+
+void Citro3DRenderer::SubmitDraw(
+	DWORD meshId,
+	const D3DRMMATRIX4D& modelViewMatrix,
+	const D3DRMMATRIX4D& worldMatrix,
+	const D3DRMMATRIX4D& viewMatrix,
+	const Matrix3x3& normalMatrix,
+	const Appearance& appearance
+)
+{
+	C3D_Mtx modelView;
+	ConvertMatrix(modelViewMatrix, &modelView);
+	C3D_FVUnifMtx4x4(GPU_VERTEX_SHADER, uLoc_modelView, &modelView);
+
+	auto& mesh = m_meshs[meshId];
+
+	C3D_BufInfo* bufInfo = C3D_GetBufInfo();
+	BufInfo_Init(bufInfo);
+	BufInfo_Add(bufInfo, mesh.vbo, sizeof(D3DRMVERTEX), 3, 0x210);
+
+	C3D_FVUnifSet(
+		GPU_VERTEX_SHADER,
+		uLoc_meshColor,
+		appearance.color.r / 255.0f,
+		appearance.color.g / 255.0f,
+		appearance.color.b / 255.0f,
+		appearance.color.a / 255.0f
+	);
+
+	C3D_FVUnifSet(GPU_VERTEX_SHADER, uLoc_shininess, appearance.shininess / 255.0f, 0.0f, 0.0f, 0.0f);
+
+	if (appearance.textureId != NO_TEXTURE_ID) {
+		C3D_TexBind(0, &m_textures[appearance.textureId].c3dTex);
+		C3D_TexEnv* env = C3D_GetTexEnv(0);
+		C3D_TexEnvInit(env);
+		C3D_TexEnvSrc(env, C3D_Both, GPU_TEXTURE0, GPU_PRIMARY_COLOR, GPU_PRIMARY_COLOR);
+		C3D_TexEnvFunc(env, C3D_Both, GPU_MODULATE);
+	}
+	else {
+		C3D_TexBind(0, nullptr);
+		C3D_TexEnv* env = C3D_GetTexEnv(0);
+		C3D_TexEnvInit(env);
+		C3D_TexEnvSrc(env, C3D_Both, GPU_PRIMARY_COLOR, GPU_PRIMARY_COLOR, GPU_PRIMARY_COLOR);
+		C3D_TexEnvFunc(env, C3D_Both, GPU_REPLACE);
+	}
+
+	C3D_DrawArrays(GPU_TRIANGLES, 0, mesh.vertexCount);
+}
+
+HRESULT Citro3DRenderer::FinalizeFrame()
+{
+	return S_OK;
+}
+
+void Citro3DRenderer::Resize(int width, int height, const ViewportTransform& viewportTransform)
+{
+	m_width = width;
+	m_height = height;
+	m_viewportTransform = viewportTransform;
+}
+
+void Citro3DRenderer::Clear(float r, float g, float b)
+{
+	StartFrame();
+	u32 color =
+		(static_cast<u32>(r * 255) << 24) | (static_cast<u32>(g * 255) << 16) | (static_cast<u32>(b * 255) << 8) | 255;
+	C3D_RenderTargetClear(m_renderTarget, C3D_CLEAR_ALL, color, 0);
+}
+
+void Citro3DRenderer::Flip()
+{
+	C3D_FrameEnd(0);
+	gfxFlushBuffers();
+	gspWaitForVBlank();
```

Comment:
```suggestion
	gspWaitForVBlank();
	g_rendering = false;
```

Reply:
but then where would C3D_FrameEnd get called? (I'm testing this right now)

Reply:
yeah this just throws out all drawing, black screen only and then it hits a svc breakpoint